### PR TITLE
elasticsearch activity tracking

### PR DESCRIPTION
**Pull request by richardstartin**
Traces all suspends/resumes in elasticsearch 7.3 `TransportClient` e.g. for a get query:

Before:

```
Activity checkpoints by thread ordered by time
Test worker:                                     |-startSpan/1-|-startSpan/2-|-suspend/2-|----------|-----------|-----------|-endSpan/1-|-startSpan/3-|-suspend/3-|----------|-----------|-----------|-startSpan/4-|-suspend/4-|----------|-----------|-----------|-startSpan/5-|-suspend/5-|----------|-----------|-------------|-----------|-----------|-startSpan/7-|-suspend/7-|----------|-----------|-----------|-startSpan/8-|-suspend/8-|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#11]: |-------------|-------------|-----------|-resume/2-|-endTask/2-|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-------------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|
elasticsearch[_client_][listener][T#1]:          |-------------|-------------|-----------|----------|-----------|-endSpan/2-|-----------|-------------|-----------|----------|-----------|-endSpan/3-|-------------|-----------|----------|-----------|-endSpan/4-|-------------|-----------|----------|-----------|-------------|-----------|-endSpan/5-|-------------|-----------|----------|-----------|-endSpan/7-|-------------|-----------|----------|-----------|-endSpan/8-|
elasticsearch[_client_][transport_worker][T#12]: |-------------|-------------|-----------|----------|-----------|-----------|-----------|-------------|-----------|-resume/3-|-endTask/3-|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-------------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#13]: |-------------|-------------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|-resume/4-|-endTask/4-|-----------|-------------|-----------|----------|-----------|-------------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#14]: |-------------|-------------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|-resume/5-|-endTask/5-|-------------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|
elasticsearch[test-node][write][T#1]:            |-------------|-------------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-startSpan/6-|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|
elasticsearch[test-node][generic][T#9]:          |-------------|-------------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-------------|-endSpan/6-|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#9]:  |-------------|-------------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-------------|-----------|-----------|-------------|-----------|-resume/7-|-endTask/7-|-----------|-------------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#10]: |-------------|-------------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-------------|-----------|-----------|-------------|-----------|----------|-----------|-----------|-------------|-----------|-resume/8-|-endTask/8-|-----------|

```

After:

```
Activity checkpoints by thread ordered by time
Test worker:                                     |-startSpan/1-|-startSpan/2-|-suspend/2-|-suspend/2-|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-endSpan/1-|-startSpan/3-|-suspend/3-|-suspend/3-|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-startSpan/4-|-suspend/4-|-suspend/4-|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-startSpan/5-|-suspend/5-|-suspend/5-|----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-startSpan/7-|-suspend/7-|-suspend/7-|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-startSpan/8-|-suspend/8-|-suspend/8-|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#11]: |-------------|-------------|-----------|-----------|-resume/2-|-endTask/2-|-resume/2-|-suspend/2-|-endTask/2-|----------|-----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][listener][T#1]:          |-------------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|-resume/2-|-endSpan/2-|-endTask/2-|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|-resume/3-|-endSpan/3-|-endTask/3-|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|-resume/4-|-endSpan/4-|-endTask/4-|-------------|-----------|-----------|----------|-----------|-------------|-----------|----------|-----------|-----------|-resume/5-|-endSpan/5-|-endTask/5-|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|-resume/7-|-endSpan/7-|-endTask/7-|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|-resume/8-|-endSpan/8-|-endTask/8-|
elasticsearch[_client_][transport_worker][T#12]: |-------------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|-----------|-resume/3-|-endTask/3-|-resume/3-|-suspend/3-|-endTask/3-|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#13]: |-------------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|-resume/4-|-endTask/4-|-resume/4-|-suspend/4-|-endTask/4-|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#14]: |-------------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|-resume/5-|-endTask/5-|-------------|-----------|-resume/5-|-suspend/5-|-endTask/5-|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[test-node][write][T#1]:            |-------------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|-startSpan/6-|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[test-node][generic][T#4]:          |-------------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|-------------|-endSpan/6-|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#9]:  |-------------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|-resume/7-|-endTask/7-|-resume/7-|-suspend/7-|-endTask/7-|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#10]: |-------------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|----------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-----------|-resume/8-|-endTask/8-|-resume/8-|-suspend/8-|-endTask/8-|----------|-----------|-----------|
```

The first activations on threads like `|-resume/2-|-endTask/2-|` come from netty and could safely be removed, they don't span any interesting activity.

### CODE REVIEW

**Comment by richardstartin** [comment]
Excluding `sendMessage` in the third commit (which can be dropped) we get:

```
Activity checkpoints by thread ordered by time
Test worker:                                     |-startSpan/1-|-startSpan/2-|-suspend/2-|----------|-----------|-----------|----------|-----------|-----------|-endSpan/1-|-startSpan/3-|-suspend/3-|----------|-----------|-----------|----------|-----------|-----------|-startSpan/4-|-suspend/4-|----------|-----------|-----------|----------|-----------|-----------|-startSpan/5-|-suspend/5-|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-startSpan/7-|-suspend/7-|----------|-----------|-----------|----------|-----------|-----------|-startSpan/8-|-suspend/8-|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#11]: |-------------|-------------|-----------|-resume/2-|-suspend/2-|-endTask/2-|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][listener][T#1]:          |-------------|-------------|-----------|----------|-----------|-----------|-resume/2-|-endSpan/2-|-endTask/2-|-----------|-------------|-----------|----------|-----------|-----------|-resume/3-|-endSpan/3-|-endTask/3-|-------------|-----------|----------|-----------|-----------|-resume/4-|-endSpan/4-|-endTask/4-|-------------|-----------|-------------|-----------|----------|-----------|-----------|-resume/5-|-endSpan/5-|-endTask/5-|-------------|-----------|----------|-----------|-----------|-resume/7-|-endSpan/7-|-endTask/7-|-------------|-----------|----------|-----------|-----------|-resume/8-|-endSpan/8-|-endTask/8-|
elasticsearch[_client_][transport_worker][T#12]: |-------------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|-resume/3-|-suspend/3-|-endTask/3-|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#13]: |-------------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-resume/4-|-suspend/4-|-endTask/4-|----------|-----------|-----------|-------------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[test-node][write][T#1]:            |-------------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-startSpan/6-|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[test-node][generic][T#4]:          |-------------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-------------|-endSpan/6-|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#14]: |-------------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-------------|-----------|-resume/5-|-suspend/5-|-endTask/5-|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#9]:  |-------------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-resume/7-|-suspend/7-|-endTask/7-|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[_client_][transport_worker][T#10]: |-------------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-resume/8-|-suspend/8-|-endTask/8-|----------|-----------|-----------|
```

**Comment by richardstartin** [comment]
@luhenry The muzzle failure is interesting:https://app.circleci.com/pipelines/github/DataDog/dd-trace-java/9458/workflows/ba435b51-8c5f-4b80-bcfa-a4454ffafdde/jobs/181787

It is saying that this solution is more general than elasticsearch 7.3 so the instrumentation can be moved (i.e. less work to do)

File: dd-java-agent/instrumentation/elasticsearch/transport/src/main/java/datadog/trace/instrumentation/elasticsearch/ThreadedActionListenerInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package datadog.trace.instrumentation.elasticsearch7_3;
```
Comment (devinsba) [inline dd-java-agent/instrumentation/elasticsearch/transport/src/main/java/datadog/trace/instrumentation/elasticsearch/ThreadedActionListenerInstrumentation.java:1]: Looks like your refactor missed the package name

**Review by devinsba — APPROVED** [review]

File: dd-java-agent/instrumentation/elasticsearch/transport/src/main/java/datadog/trace/instrumentation/elasticsearch/ThreadedActionListenerInstrumentation.java (hunk 1 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/elasticsearch/transport/src/main/java/datadog/trace/instrumentation/elasticsearch/ThreadedActionListenerInstrumentation.java]: Thanks! The build would have let me know eventually 😆

**Comment by jbachorik** [comment]
This looks a bit odd
```elasticsearch[test-node][write][T#1]:            |-------------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-startSpan/6-|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|
elasticsearch[test-node][generic][T#4]:          |-------------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|-------------|-endSpan/6-|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|-------------|-----------|----------|-----------|-----------|----------|-----------|-----------|
```

Starting a span on one thread and finishing it on another without any migration.

**Comment by richardstartin** [comment]
@jbachorik The span which doesn't migrate here is an cluster admin operation which we shouldn't be tracing anyway, it bypasses this mechanism because it is a `BaseFuture`, which this driver reserves for very lightweight operations. I would prefer to suppress the span than support this particular case.

**Comment by richardstartin** [comment]
Concretely, the span not migrated relates to a `PutMappingRequest`/`MappingUpdatedAction` and the span wraps this method:
```java

    public void updateMappingOnMaster(Index index, String type, Mapping mappingUpdate, ActionListener<Void> listener) {
        if (type.equals(MapperService.DEFAULT_MAPPING)) {
            throw new IllegalArgumentException("_default_ mapping should not be updated");
        }
        client.preparePutMapping().setConcreteIndex(index).setType(type).setSource(mappingUpdate.toString(), XContentType.JSON)
            .setMasterNodeTimeout(dynamicMappingUpdateTimeout).setTimeout(TimeValue.ZERO)
            .execute(new ActionListener<AcknowledgedResponse>() {
                @Override
                public void onResponse(AcknowledgedResponse acknowledgedResponse) {
                    listener.onResponse(null);
                }

                @Override
                public void onFailure(Exception e) {
                    listener.onFailure(unwrapException(e));
                }
            });
    }
```
which has nothing to do with user activity and is probably confusing. I'm not sure if this would happen once per request or if it is a scheduled chore, but I feel it should not be traced. Would it complicate matters to leave this in place and allow the span not to be migrated? Then it can be deferred to later to decide whether this action should even be traced.

**Comment by jbachorik** [comment]
> Would it complicate matters to leave this in place and allow the span not to be migrated?

Following the current parsing rules and the fact that we need to account for profiles containing only partial information (the trace boundaries do not align with the recording boundaries) leaving it in the current state would basically mean that the span would be considered active on T#1 until the end of the recording. Similarly, on T#4 we would have to assume that this span has been active since the beginning of the recording because we have no other information (it might have been started/resumed in the previous recording - we have no way to tell).

**Comment by richardstartin** [comment]
@jbachorik I added logic to migrate `PutMappingRequest` spans manually to ensure there is a balance until it's decided whether to trace these actions or not.

**Review by jbachorik — APPROVED** [review]

**Pull request merged**